### DISCUSSION TITLE
minimal-gas-prices and baseFeePerGas conflicts

### DISCUSSION THREAD

**Discussion by yihuang** [tharsis/ethermint#830] [General]
Currently, when the feemarket module is enabled, for eth transactions we have an extra requirement:
```
effectiveGasPrice >= minimal-gas-prices
where effectiveGasPrice = min(baseFeePerGas + tipCap, feeCap) for `DynamicFeeTx`
      effectiveGasPrice = gasPrice for `LegacyTx`
```

Usually, clients pick zero or constant `tipCap` for dynamic fee tx, and `baseFeePerGas` keep dropping when blocks are sparse, when `baseFeePerGas` drop to a point where `effectiveGasPrice < minimal-gas-prices`, the dynamic fee tx will fail, and client have to re-calculate a `tipCap` based on current `minimal-gas-prices` and `baseFeePerGas`. So basically the client logic will be different from go-ethereum.

A simple solution is don't check `minimal-gas-prices` at all for eth tx when feemarket is enabled. And for legacy tx, rather than check `tx.gasPrice >= minimal-gas-prices`, we check `tx.gasPrice >= baseFeePerGas`.

And considering the native cosmos tx, we shouldn't set min-gas-price too low, a reasonable one might be set `min-gas-price = initial base fee`.

**Comment by thomas-nguy** [comment]
isnt `tipCap` = `minimal-gas-prices`?

in that case effectiveGasPrice can never be inferior to `minimal-gas-price`?

(in case feecap is lower, then its normal for tx to fail)

**Reply by yihuang** [reply]
That'll cost `minimal-gas-prices + baseFeePerGas` for dynamic fee tx, while only cost `minimal-gas-prices` for legacy tx.

**Reply by thomas-nguy** [reply]
yeah i see, for legacy tx, if feemarket module is enable, it would be more sense to check `baseFee` instead of `min-gas-price`

**Reply by thomas-nguy** [reply]
btw what do you mean by legacy tx?  

I think we dont make the distinction on ante handler between legacy and dynamic tx and both txs should follow the same rule

- if feemarket enable then check that `txfee` >= baseFee + min-gas-price 
- if feemarket not enable then check that `txFee` >= min-gas-price

`txFee` = effectiveFee for  dynamicTx (EIP 1559) or  = gasprice for legacytx

**Reply by yihuang** [reply]
For `DynamicFeeTx`, `effectiveGasPrice` is `min(baseFeePerGas + tipCap, feeCap)`.
For `LegacyTx`, `effectiveGasPrice` is `gasPrice`.

**Reply by yihuang** [reply]
Currently, the `eth_gasPrice` API returns `baseFeePerGas + minimal-gas-price`, I think that's wrong too.

**Comment by robert-zaremba** [comment]
In your model, is `minimal-gas-price` part of consensus or a block producer parameter?

**Reply by yihuang** [reply]
> In your model, is `minimal-gas-price` part of consensus or a block producer parameter?

block producer parameter, what we have in cosmos-sdk currently.
